website/blog/2020-06-16-alluxio-datacaching.md

```diff
@@ -0,0 +1,244 @@
+It is critical for the coordinator to leverage the same worker for a split which may contain the data for it in its cache.
+<br>
+
+![](/img/blog/2020-06-16-alluxio-datacaching/presto_affinity_scheduler.jpg)
+<br>
+
+The above diagram illustrates how affinity scheduling distributes various splits among the workers.
+<br><br>
+
+Soft affinity scheduling makes the best attempt to assign the same split to the same worker when doing the scheduling. The soft affinity scheduler uses the hash of a split to 
+choose a preferred worker for the split. Soft affinity scheduler:
+1. Computes a preferred worker for a spit. If the preferred worker has resources available then it is assigned the split.
+2. If the preferred worker is busy then the coordinator chooses a secondary preferred worker, and assigns the split if resources are available.
+3. If the secondary preferred worker is also busy then the coordinator assigns the split to the least busy worker.
+
+![](/img/blog/2020-06-16-alluxio-datacaching/presto_affinity_scheduler_algo.jpg)
+
+The definition of a busy node is defined by two configs: 
+1. Max splits per node:  *node-scheduler.max-splits-per-node*
+2. Max pending splits per task: *node-scheduler.max-pending-splits-per-task*
+
+Once the number of splits on one node exceeds one of the above configured limitations, this node would be treated as a busy node.
+<br><br>
+As it can be observed, node affinity is absolutely critical for cache effectiveness. Without node affinity, the same split may be processed by 
+different workers at different times, which can make caching the split data redundant. 
+<br><br>
+Due to this, if the affinity scheduler fails to assign the split to a preferred worker (because it was busy), it signals the assigned worker 
+to not cache the split data. It means the worker would only cache the split data if it is the primary or secondary preferred worker for the split. 
+
+## Alluxio data cache
+
+Alluxio file system is an open-source data orchestration system that is often used as a distributed caching service to Presto. 
+To achieve sub-second query latencies in our architecture, we want to further reduce the communication overhead between Presto and Alluxio. 
+As a result, core teams from Alluxio and Presto collaborated to carve out a single-node, embedded cache library from the Alluxio service.
+<br><br>
+In particular, a Presto worker queries this Alluxio local cache inside the same JVM through a standard HDFS interface. 
+On a cache hit, Alluxio local cache directly reads data from the local disk and returns the cached data to Presto; 
+otherwise, it retrieves data from the remote data source, and caches the data on the local disk for followup queries. 
+This cache is completely transparent to Presto. In case the cache runs into issues (e.g., local disk failures), the Presto reads 
+fallback to the remote data source. This workflow is shown as the figure below.
```

Comment:
fallback is a noun, the verb is fall back. https://english.stackexchange.com/questions/90549/difference-between-fallback-and-fall-back